### PR TITLE
Rework `.spec.kubernetes.kubelet.systemReserved` deprecation in `Shoot` API

### PR DESCRIPTION
**Pull request by MichaelEischer**
**How to categorize this PR?**

/area control-plane
/kind enhancement

**What this PR does / why we need it**:

Reworks #9985 to allow for a smoother deprecation of the `systemReserved` field from `shoot.Spec.Kubernetes.Kubelet` (and the worker-specific counterpart).
Now, the WorkerPoolHashV2 includes "kubeReserved+systemReserved", such that a node-pool roll only happens if their sum changes. This allows for migrating away from `systemReserved` even once the `NewWorkerPoolHash` feature gate is enabled.

The WorkerPoolHashV2 will change once for shoots where `systemReserved` is not empty, however, this should be fine as the feature gate is still alpha and most provider extensions don't yet support it.

/cc @rfranzke @timuthy

**Which issue(s) this PR fixes**:

Part of #2590

**Special notes for your reviewer**:

Reworks #9985, see above.

- I'm not entirely sure what the release notes should look like.
- How to keep track of the changes require to upgrade a shoot?

**Release note**:

```breaking operator
When the `NewWorkerPoolHash` feature gate is enabled, the calculation now also rolls worker nodes of `Shoot`s when changing `systemReserved` in the `kubelet` configuration. Worker pools are not rolled if the sum of `kubeReserved` and `systemReserved` does not change. If the feature gate is already enabled, then the worker pools of `Shoot`s with non-zero values in `systemReserved` will be rolled once.
```

### CODE REVIEW

File: plugin/pkg/shoot/validator/admission.go [hunk 1]
```diff
@@ -1368,6 +1369,11 @@ func validateKubeletConfig(fldPath *field.Path, machineTypes []gardencorev1beta1
 		if kubeletConfig.SystemReserved.Memory != nil {
 			reservedMemory.Add(*kubeletConfig.SystemReserved.Memory)
 		}
+		k8sGreaterOrEqual131, _ := semver.NewConstraint(">= 1.31")
+
+		if kubeletVersion != nil && k8sGreaterOrEqual131.Check(kubeletVersion) {
```
Comment (rfranzke) [inline plugin/pkg/shoot/validator/admission.go]: Why did you put this here? I would have expected to see this in the "static validation" in https://github.com/gardener/gardener/blob/a68e11ff6d03f80f0aaff4f4c225255c1fe93ceb/pkg/apis/core/validation/shoot.go#L1788-L1880

File: pkg/controllermanager/controller/shoot/maintenance/reconciler.go [hunk 2]
```diff
@@ -224,6 +224,10 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gard
 		}
 	}
 
+	// TODO(MichaelEischer): Kubernetes.Kubelet.SystemReserved is no longer supported in Gardener starting from K8S 1.31.
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/maintenance/reconciler.go]:
```suggestion
	// TODO(ialidzhikov): Kubernetes.Kubelet.SystemReserved is no longer supported in Gardener starting from K8S 1.31.
```

ref https://github.com/gardener/gardener/issues/10286#issuecomment-2275673091

**Review by rfranzke — CHANGES_REQUESTED** [review]
Please change

- https://github.com/gardener/gardener/blob/a68e11ff6d03f80f0aaff4f4c225255c1fe93ceb/pkg/apis/core/types_shoot.go#L908
- https://github.com/gardener/gardener/blob/a68e11ff6d03f80f0aaff4f4c225255c1fe93ceb/pkg/apis/core/v1beta1/types_shoot.go#L1179

(and run `make generate`) to remove this field only when support for Kubernetes 1.31 is dropped.

Q: Do we have to wait for 1.31 now in https://github.com/gardener/gardener/blob/a68e11ff6d03f80f0aaff4f4c225255c1fe93ceb/pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go#L943?

**Comment by rfranzke** [comment]
> I'm not entirely sure what the release notes should look like.

Why not? Do you have ideas? I guess we should talk about the changes related to the `NewWorkerPoolHash` feature gate?

Information about the fact that `systemReserved` cannot be set for `1.31` shoots should probably be published as part of https://github.com/gardener/gardener/issues/10286

File: plugin/pkg/shoot/validator/admission.go (hunk 1 above)
Comment (MichaelEischer) [reply plugin/pkg/shoot/validator/admission.go]: Oh, I've missed that file. Thanks! The validation is now moved there.

File: pkg/controllermanager/controller/shoot/maintenance/reconciler.go (hunk 2 above)
Comment (MichaelEischer) [reply pkg/controllermanager/controller/shoot/maintenance/reconciler.go]: I've updated the comment.

**Comment by MichaelEischer** [comment]
> to remove this field only when support for Kubernetes 1.31 is dropped.

Do we want to remove support for it after K8S 1.30 or 1.31 is dropped? Once there are no more clusters with K8S 1.30, then the field should be unused. So it can already be removed at that point, or do we need more overlap?

> I guess we should talk about the changes related to the `NewWorkerPoolHash` feature gate?

I've updated the changelog entry. Right now, I've marked it as "breaking" given that the change might have an impact on shoots that already use the feature gate.

**Comment by rfranzke** [comment]
> Do we want to remove support for it after K8S 1.30 or 1.31 is dropped? Once there are no more clusters with K8S 1.30, then the field should be unused. So it can already be removed at that point, or do we need more overlap?

You're right, we can remove it once we drop support for 1.30 :) Thanks!

**Review by rfranzke — COMMENTED** [review]
Very nice @MichaelEischer 

/lgtm
/approve

**Pull request merged**